Add follower synchronization mechanism

ClearlyClaire opened the pull request:
Fixes #14480

Everything in here is an ActivityPub extension that is not exactly in ActivityPub's spirit (since it's domain-based, and introduces synchronization mechanisms that are not a thing in AP to my knowledge) but consistent with the assumptions implementations already make (remote actors being able to create objects with any identifier as long as it is on the same domain name as the actor itself, for instance).

While the PR should be in good shape, everything, especially the vocabulary, may change from feedback from other implementors, as the issue is not strictly a Mastodon issue, but a more general one.

Related discussion: https://socialhub.activitypub.rocks/t/proposal-mechanism-for-synchronizing-followers-lists-across-servers/852

## Context

There are many ways—such as software bugs (e.g. the one fixed by #14479, Pleroma forgetting about followers a while ago, and probably other bugs we did not catch or may happen in the future), incompatibilities (while initiating and accepting a follow is pretty well-defined in AP, forcefully removing followers is not), prolonged downtimes, instances being brought back from the dead or restored from backups—in which two instances A and B, acting in good faith, may not agree with who from A follows someone on B.

This may cause confusion and frustration as one may think they are followed/follow by someone they are not. Worse: with Mastodon’s follower-only privacy scope, this can get particularly damaging, as receiving remote instances are asked to deliver the message to who they know to follow the poster: if that gets out of sync, then the messages may reach people that the poster expects to not have access to.

Therefore, a mechanism to detect synchronization issues and correct them is needed. Such a mechanism has to take into account that the full list of an account’s followers may not be public information, and should thus only consider followers residing on a specific domain.

## Proposal

I propose the addition of an optional `Collection-Synchronization` HTTP header following the same syntax as the [Signature header](https://tools.ietf.org/html/draft-ietf-httpbis-message-signatures-00.html#section-4), with the following attributes:
- `collectionId`: must be the sender's `followers` collection
- `url`: an URL to a partial collection containing the identifiers of the sender's followers residing on the receiver's instance. MUST reside on the same domain as the actor itself, and SHOULD be only accessible with a signed query from the receiver's instance
- `digest`: hexadecimal representation of the XORed SHA256 digests of each of the identifiers in the partial collection

### Example

When delivering a private message to `https://mastodon.social/users/foo/inbox`, the instance running on `social.sitedethib.com` sets the following HTTP header:
`Collection-Synchronization: collectionId="https://social.sitedethib.com/users/Thib/followers", url="https://social.sitedethib.com/users/Thib/followers_synchronization", digest="b08ab6951c7d6cc2b91e17ebd9557da7fae02489728e9332fcb3a97748244d50"`

The result of querying `https://social.sitedethib.com/users/Thib/followers_synchronization` (when signing the request on behalf of a mastodon.social account):
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "https://social.sitedethib.com/users/Thib/followers?domain=mastodon.social",
  "type": "OrderedCollection",
  "orderedItems": [
    "https://mastodon.social/users/Gargron"
  ]
}
```

## FAQ

#### Why `partialCollection`, why not re-use the followers collection?

Currently, Mastodon lists either all followers in the `followers collection`, or none of them (if the user has opted in to “hide their network”). The first behavior is not suitable for synchronization, because it would be potentially much larger than needed. The second would obviously not work.

The behavior we want from `partialCollection`, that is, listing all known followers residing on the requesting instance, is different from those two options. While we could switch to that behavior when receiving a signed request:
1. Implementations may still want a chance to get the full list of followers when signing a request (e.g., implementations allowing access to the full list of followers, but only to followers)
2. Mastodon currently signs all outgoing requests, and uses the fact that the collection cannot be paginated to respect the “hide your network” option across instances.

#### Wouldn't that leak followers information?

No, if the `partialCollection` requires an HTTP Signature, and only returns followers residing on that instance, it will only return information that the requester is supposed to know about (but might not because of bugs or system failures).

Also, the `digest` is only delivered to the `inbox` via an HTTP header, unaware applications will simply ignore it, and will not risk relaying it anywhere.

## Implementation notes

- **caching hashes**
  Hashes are cached in Rails' cache, and the cache gets invalidated when followers are created or removed.
  The cache key used is `followers_hash:#{account_id}:#{domain}` where `domain` is the remote domain we are interested in if the account is local, or `local` if it's a remote account (and we thus are interested in the local followers).
- **followers synchronization endpoint**
  The followers synchronization endpoint provided by this PR is separate from the regular followers collection endpoint, and does not expect a parameter for the requested domain. Indeed, a parameter would be redundant as the endpoint *requires* a signed fetch, and the used domain is retrieved from that signature.
  In Mastodon's case, having a separate endpoint actually makes the code less complex and easier to reason about.

## TODO

- [x] Implement receiving end
- [x] Add tests for receiving end
- [x] Implement sending end
- [x] Add tests for sending end

ClearlyClaire committed:
Add support for followers synchronization on the receiving end

Check the `collectionSynchronization` attribute on `Create` and `Announce`
activities and synchronize followers from provided collection if possible.

ClearlyClaire committed:
Add tests for followers synchronization on the receiving end

cwebber commented:
This seems to assume that a complete list of who I'm following is public information, whereas it might only be a partial subset?  This could be a privacy problem?

Unless it's a hash only for that domain.  That could work, though it makes me unhappy because I don't like the assumption of ActivityPub being a domain-oriented system, which I think will create problems long term...

ThibG commented:
The `partialCollection` would only contain what the remote domain (identified by `domain`) is supposed to know, and the `hash` is indeed specific to it.

I understand your concern, but it is precisely in the context of domain-oriented systems where the issue arises.

cwebber commented:
In general I think that a "do I really have the correct state" synchronization system is fine and welcome; I wonder if it can be more general though than just related to followers.  The hash part at least is elegant and a good idea.  For that component, maybe even better would be to use a bloom filter, so that if a mismatching result is found, one could at least query to see if certain elements are *part of* the set.

I do wonder if there's a more generalized pattern; I've thought for a while that there's something to [the Unum pattern](http://habitatchronicles.com/2019/08/the-unum-pattern/) that could be followed.

cwebber commented:
Yes, I realize I'm pushing more abstract patterns for people who are doing more practical "AP as deployed" approaches (a legitimate criticism).

The domain approach could be generalized though: instead of domain, you could have "node"; conceptually, each Actor could list on its profile the "Node" actor that it's associated with corresponding to its hub in the hub-and-spoke model.

This then could be the actual node that you would query to see whether or not your followers were in that set.

The main problem I guess with that approach is to make it secure you need to have each Node/Hub sign that the actor really does belong to that node/hub.

ThibG commented:
> For that component, maybe even better would be to use a bloom filter, so that if a mismatching result is found, one could at least query to see if certain elements are _part of_ the set.

I have thought of that. It might be a good idea. The workflow with this would be, if there's a mismatch, use the provided bloom filter to prune local followers, re-compute the local bloom filter, and if it still doesn't match, perform the HTTP query. This would spare an HTTP query in many (most?) cases, but I am unsure it is worth the computational or code complexity (I'll have a look).

> I do wonder if there's a more generalized pattern; I've thought for a while that there's something to [the Unum pattern](http://habitatchronicles.com/2019/08/the-unum-pattern/) that could be followed.

I will have a look.



> The domain approach could be generalized though: instead of domain, you could have "node"; conceptually, each Actor could list on its profile the "Node" actor that it's associated with corresponding to its hub in the hub-and-spoke model.

I'm afraid a lot of security assumptions are already made across most if not all ActivityPub software around the fact that two URLs behind a same domain are managed by the same logical entity. Introducing a concept of managing “Node” might be too late.

ClearlyClaire committed:
Add support for follower synchronization on the sender's end

ClearlyClaire committed:
Add tests for the sending end

ThibG commented:
Alright, I think I have roughly finished a first draft of that feature.

Some notes:
- while a bloom filter could be useful, that's probably overkill, the feature isn't supposed to kick in often anyway. And since the same algorithm with the same parameters would need to be used across implementations, that is probably not worth it.
- I guess we can consider the collection an Unum, but I am not sure what that gives us. I don't believe the article linked earlier really delves into anything useful for the case at hand.
- There is something I'm a bit worried out: different implementations considering different “canonical” domain names for accounts (e.g. when `LOCAL_DOMAIN` ≠ `WEB_DOMAIN` in Mastodon). I have a hunch this isn't an issue unless some implementation considers accounts from the same logical instance under different domain names, but I'm too tired right now to think more about this.

ThibG commented:
Re-using existing ActivityStreams vocabulary:
- `collection` could be `subject` (in AS, only used for relationships) or `describes` (in AS, only used for profiles), simply `object`, or even `partOf` (used for Collections pages)
- `partialCollection` could be `url`
- I can't think anything that would be suitable for `domain`
- as far as the `hash` is concerned, it probably would make sense to use https://w3c-ccg.github.io/security-vocab/#Digest even if it is more verbose

lanodan commented:
Sounds like a good proposal to me, few small things though:

- `domain`: Isn't it actually the hostname? (and it's the one from the ActivityPub id not webfinger, right?)
- `hash`: I'm not sure about putting it there, maybe a last-modified date would be better, also SHA256 is fine right now but might not be in some years.

Also why not use authenticated fetching (HTTP Signatures) so that the `followers` endpoint just keeps working and is refreshed periodically, like the actor currently is in pleroma and mastodon?

ThibG commented:
> Sounds like a good proposal to me, few small things though:
> 
>     * `domain`: Isn't it actually the hostname? (and it's the one from the ActivityPub id not webfinger, right?)

It's the domain part of the handle, so it's the webfinger thing. I have to think a bit more whether that would cause issues with software that use the AP hostname instead. My hunch is that it wouldn't cause issues as long as:
1. On the receiving end, when checking the `domain`, they accept both the webfinger thing and the hostname of AP endpoints (that's what my implementation does)
2. On the sending end, when computing the domain-specific hash, all of the remote followers on the remote instance have to be considered with the same domain. Meaning that two accounts on `domain.org` served by `social.domain.org` need to be referred both as `domain.org` or `social.domain.org`, not a mix of that. Mastodon will always use the webfinger name, and I think Pleroma will always use the AP hostname, but I am not completely sure about it.

>     * `hash`: I'm not sure about putting it there, maybe a last-modified date would be better, also SHA256 is fine right now but might not be in some years.

I think something like Last-modified would be kinda error-prone, plus you have to store it and can never re-calculate it.

About SHA256, I guess that would be solved by using https://w3c-ccg.github.io/security-vocab/#Digest as I suggested in my last post.

> Also why not use authenticated fetching (HTTP Signatures) so that the `followers` endpoint just keeps working and is refreshed periodically, like the actor currently is in pleroma and mastodon?

Depending on the user's settings, the followers endpoint can currently list:
- all of the followers
- none of the followers

Having a third option (only the followers from the requesting domain) tied to HTTP Signatures does not make a lot of sense (you may want to know all the followers). Plus, currently, Mastodon figures out whether a remote user wants to show their followers by checking if they can access that collection's first page. That assumption would be broken by having that third option.

Furthermore, periodic profile checking in Mastodon is actually not done for remote accounts which are followed: indeed, Mastodon relies on receiving `Update`s to get the up-to-date info and avoid spurious fetches. One may argue that this poses a similar synchronization issue, though, although one that doesn't put privacy at risk.

In my implementation, I found that using a separate endpoint was simply easier.

ThibG commented:
Pushed a few commits to change the ActivityStreams vocabulary used, that is:
- `object` instead of `collection`
- `domain` is kept but pulled from the w3c security context
- `hash` is replaced with `digest` from the w3c security context, and instead of having a base64-encoded SHA256 digest, contains a `Digest` object from the w3c security context (with `http://www.w3.org/2001/04/xmlenc#sha256` being the only acceptable algorithm for Mastodon in the current PR)

If people agree about these changes, I will clean up the commits and rework the original description.

### New Example

```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    "https://w3id.org/security/v1",
    {
      "toot":"http://joinmastodon.org/ns#",
      "collectionSynchronization":"toot:collectionSynchronization",
      "SynchronizationItem":"toot:SynchronizationItem",
      "partialCollection": "toot:partialCollection"
  }],
  "id": "https://social.sitedethib.com/users/Thib/statuses/1234",
  "type": "Note",
  "attributedTo": "https://social.sitedethib.com/users/Thib",
  "to": ["https://social.sitedethib.com/users/Thib/followers"],
  "content": "This is just a private toot",
  "collectionSynchronization": [
    {
      "type": "SynchronizationItem",
      "domain": "mastodon.social",
      "object": "https://social.sitedethib.com/users/Thib/followers",
      "partialCollection": "https://social.sitedethib.com/users/Thib/followers_synchronization",
      "digest": {
        "digestAlgorithm": "http://www.w3.org/2001/04/xmlenc#sha256",
        "digestValue": "b08ab6951c7d6cc2b91e17ebd9557da7fae02489728e9332fcb3a97748244d50"
      }
    }
  ]
}
```

nightpool commented:
the `domain` from the CCG Security Vocabulary is something like `"issuer.interop.transmute.world"`, I'm not sure it's appropriate to use here, it's a complete separate concept from a URL's domain

More broadly, would you be okay with posting this on the https://social.activitypub.rocks forum? There we can solicit more feedback from a broader range of implementors, and get more information up front.

ThibG commented:
@nightpool I think this is close enough, it's still a `domain` name, even if its purpose in the context of a `Signature` is slightly different. https://w3c-ccg.github.io/security-vocab/#proofValue for instance has a more approachable example



> More broadly, would you be okay with posting this on the https://socialhub.activitypub.rocks forum? There we can solicit more feedback from a broader range of implementors, and get more information up front.

Good point, doing that.

nightpool commented on app/serializers/activitypub/activity_serializer.rb:
```diff
@@ -12,11 +12,55 @@ def self.serializer_for(model, options)
+  end
+
   attributes :id, :type, :actor, :published, :to, :cc
+  has_many :collection_synchronization, serializer: SynchronizationItemSerializer if: :collection_synchronization?
```
need a comma here
```suggestion
  has_many :collection_synchronization, serializer: SynchronizationItemSerializer, if: :collection_synchronization?
```

ThibG replied:
yup, shuffling things around, didn't catch that!

ClearlyClaire committed:
Switch from AS attributes to HTTP header

Replace the custom `collectionSynchronization` ActivityStreams attribute by
an HTTP header (`X-AS-Collection-Synchronization`) with the same syntax as
the `Signature` header and the following fields:
- `collectionId` to specify which collection to synchronize
- `digest` for the SHA256 hex-digest of the list of followers known on the
   receiving instance (where “receiving instance” is determined by accounts
   sharing the same host name for their ActivityPub actor `id`)
- `url` of a collection that should be fetched by the instance actor

Internally, move away from the webfinger-based `domain` attribute and use
account `uri` prefix to group accounts.

ClearlyClaire committed:
Add environment variable to disable followers synchronization

Since the whole mechanism relies on some new preconditions that, in some
extremely rare cases, might not be met, add an environment variable
(DISABLE_FOLLOWERS_SYNCHRONIZATION) to disable the mechanism altogether and
avoid followers being incorrectly removed.

The current conditions are:
1. all managed accounts' actor `id` and inbox URL have the same URI scheme and
   netloc.
2. all accounts whose actor `id` or inbox URL share the same URI scheme and
   netloc as a managed account must be managed by the same Mastodon instance
   as well.

As far as Mastodon is concerned, breaking those preconditions require extensive
configuration changes in the reverse proxy and might also cause other issues.

Therefore, this environment variable provides a way out for people with highly
unusual configurations, and can be safely ignored for the overwhelming majority
of Mastodon administrators.

ClearlyClaire committed:
Only set follower synchronization header on non-public statuses

This is to avoid unnecessary computations and allow Follow-related
activities to be handled by the usual codepath instead of going through
the synchronization mechanism (otherwise, any Follow/Undo/Accept activity
would trigger the synchronization mechanism even if processing the activity
itself would be enough to re-introduce synchronization)

Gargron commented on app/controllers/activitypub/followers_synchronizations_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  private
+
+  def set_items
+    @items = @account.followers.where(domain: signed_request_account.domain).pluck(:uri).sort
```
Might potentially want to add pagination logic since someone could have more than a thousand followers from a single domain even right now.

Gargron replied:
Actually, there doesn't seem to be a good reason to use alphabetic sort here at all. When another server loads this data, it's not used to build a digest, individual items are used and order does not matter. So, using standard pagination order would ensure that the collection is deterministically paginable.

Gargron commented on app/controllers/activitypub/inboxes_controller.rb:
```diff
@@ -52,6 +53,20 @@ def upgrade_account
+  def process_collection_synchronization
+    raw_params = request.headers['X-AS-Collection-Synchronization']
+    return if raw_params.blank?
+    return if ENV['DISABLE_FOLLOWERS_SYNCHRONIZATION'] == 'true'
```
I'd unite this with the line above. Maybe extract to `disable_followers_synchronization?` method for easier testing.

ThibG replied:
Put it on one line, not sure about a different method.

Gargron commented on app/controllers/activitypub/inboxes_controller.rb:
```diff
@@ -52,6 +53,20 @@ def upgrade_account
+
+    ActivityPub::PrepareFollowersSynchronizationService.new.call(signed_request_account, params)
+  rescue Parslet::ParseFailed
+    Rails.logger.warn 'Error parsing collection synchronization header'
```
Nitpick but I would maybe write "X-AS-Collection-Synchronization" here to be more specific in the error log. Though I also don't remember how we treat errors from "bad input" in other places. Maybe you wouldn't want to fill the warn log level with stuff like this..?

ThibG replied:
I think in general, Mastodon doesn't print enough stuff in logs, but maybe that should be `debug` loglevel rather than `warn`. Updated the text itself.

Gargron commented on app/models/concerns/account_interactions.rb:
```diff
@@ -243,6 +243,15 @@ def lists_for_local_distribution
+    key = url_prefix.nil? ? 'local' : url_prefix
+
+    Rails.cache.fetch("followers_hash:#{id}:#{key}") do
+      scope = [nil, 'local'].include?(url_prefix) ? followers.where(domain: nil) : followers.where(Account.arel_table[:uri].matches(url_prefix + '%', false, true))
```
I don't super-like that you could have `nil` or `"local"` here as the argument and they mean the same thing. Ideally it should just be one special value, and it could be `nil` like it is for `domain` in other places.

There are potentially thousands of rows being returned here. Even though it's cached, this will mess with memory usage. Ideally, `find_each` should be used. Ruby's `Digest` module supports generating digests from chunks, i.e.:

```ruby
digest = Digest::SHA256.new

scope.find_each { |record| digest.update(record.foo) }

digest.hexdigest
```

The problem is of course that you want the digest to be generated from a canonical, alphabetically sorted order, and our standard pagination order is not that.

ThibG replied:
Regarding the memory consumption... yeah, that could be an issue. I'm not too sure what we can do here since the batch order and the alphabetical order aren't going to be the same... we could at least specialize the query for local VS remote followers in order to select just the required fields.

Gargron replied:
`ORDER BY accounts.username ASC` should cover both local and remote accounts, right? You only ever select either local domain or a specific domain so you don't have to worry about domain sorting. It's not super efficient because it's not an indexed sorting order on that table but it's still more efficient than loading rows into Ruby and sorting afterwards.

Gargron replied:
That being said it wouldn't solve pagination on the rendering side because when loading next page you want to guarantee that you're in the exact place you left off even as the underlying data may shift, and alphabetic sorting does not give that.

ThibG replied:
afaik, `find_each` is based on batch queries, which I think are ordered by `id` and discard any sorting you may have specified in your scope

Gargron replied:
Although I guess there is no specific reason why the rendered collection should be in the same order as the digest? The main thing is that another server could paginate over all records, at that point the digestable order doesn't matter.

Gargron replied:
Oh right. Can we... can we compute the digest in postgres..? Pluck will be a good improvement though.

ThibG replied:
Doing it efficiently in postgres doesn't seem easy either. Ended up doing it another way: XORing individual hashes. Since we are guaranteed to not have duplicates, this should be fine. Furthermore, this avoids issues were sorting would be done with different uses on different implementations/systems.

ThibG replied:
I think using XORed individual hashes solves that problem entirely (as well as the problem of defining an ordering between textual values).

Gargron commented on app/models/follow.rb:
```diff
@@ -63,4 +65,10 @@ def decrement_cache_counters
   end
+
+  def invalidate_hash_cache
+    return if account&.local? && target_account&.local?
```
I don't think they can be nil when this is executed. If they could, you'd run into a nil error in the line below.

Gargron commented on app/services/activitypub/prepare_followers_synchronization_service.rb:
```diff
@@ -0,0 +1,15 @@
+
+    return unless params['collectionId'] == @account.followers_url
+    return if invalid_origin?(params['url'])
+    return if @account.followers_hash('local') == params['digest']
```
Could be part of the guard clause above. This whole method might look better with just one guard clause, they are simple comparisons.

If the goal of supporting both `nil` and `"local"` arguments in `followers_hash` is code readability, you could create a method like `local_followers_hash` that calls `followers_hash` with `nil`.

Gargron commented on app/services/activitypub/synchronize_followers_service.rb:
```diff
@@ -0,0 +1,63 @@
+      next if expected_follower.following?(@account)
+
+      follow = Follow.new(account: expected_follower, target_account: @account)
+      # TODO: the follow doesn't have an uri, how does that work?
```
Indeed, this is awkward. But I'm not sure why this is a TODO and a question. Perhaps change it to a comment that warns:

> Since we were not aware of the follow from our side, we do not have an ID for it that we can include in the Undo activity. For this reason, the Undo may not work with software that relies exclusively on matching activity IDs and not the actor and target

Gargron commented on app/services/activitypub/synchronize_followers_service.rb:
```diff
@@ -0,0 +1,63 @@
+    return if items.nil?
+
+    @expected_followers = items.map { |uri| ActivityPub::TagManager.instance.uri_to_resource(uri, Account) }.compact
+    # TODO: what to do with unresolvable accounts?
```
It should not be possible for "our" accounts to be unfindable since the `Account` record is always kept

ThibG replied:
Yeah. Well, we do in some cases allow removing `Account`, I think? But in any case, the case should be exceedingly rare, and if it occurs there's nothing much we can do since we don't have an Account to sign an `Undo` with. Updated the comment.

Gargron commented on app/workers/activitypub/delivery_worker.rb:
```diff
@@ -38,9 +39,15 @@ def build_request(http_client)
   end
 
+  def synchronization_header
+    url_prefix = @inbox_url[/http(s?):\/\/[^\/]+\//]
```
I guess this is the `inbox_url_prefix` and could be extracted to a method as such

Gargron commented on app/workers/activitypub/followers_synchronization_worker.rb:
```diff
@@ -0,0 +1,14 @@
+    @account = Account.find_by(id: account_id)
+    return true if @account.nil?
+
+    ActivityPub::SynchronizeFollowersService.new.call(@account, url)
```
Could be compacted to one line by doing `Account.find` and rescuing the error like we do in all other workers.

ThibG replied:
This is a pattern I do not like that much as it could silently hide unexpected `ActiveRecord::RecordNotFound` exceptions.

Gargron commented on app/controllers/activitypub/inboxes_controller.rb:
```diff
@@ -52,6 +53,20 @@ def upgrade_account
   end
 
+  def process_collection_synchronization
+    raw_params = request.headers['X-AS-Collection-Synchronization']
```
Don't want to bikeshed but my suggestion would be to save the few bytes and just name it `Collection-Synchronization`. It does not conflict with any standard headers.

ClearlyClaire committed:
Change how ActivityPub::SynchronizeFollowersService handles follow requests

If the remote lists a local follower which we only know has sent a follow
request, consider the follow request as accepted instead of sending an Undo.

ClearlyClaire committed:
Integrate review feeback

- rename X-AS-Collection-Synchronization to Collection-Synchronization
- various minor refactoring and code style changes

ClearlyClaire committed:
Only select required fields when computing followers_hash

ClearlyClaire committed:
Use actor URI rather than webfinger domain in synchronization endpoint

ClearlyClaire committed:
Change hash computation to be a XOR of individual hashes

Makes it much easier to be memory-efficient, and avoid sorting discrepancy issues.

ClearlyClaire committed:
Marginally improve followers_hash computation speed

ClearlyClaire committed:
Further improve hash computation performances by using pluck_each

tocharomera commented:
Hello👋

how could this be triggered in a self-hosted Mastodon server for a specific account?

A mastodon server database I managed got corrupted and I lost some followers on the recovery process on some accounts, now they still follow the accounts on their profiles but they dont show up on the accounts followers section of my server.

Could you give me an advise how to recover them?
Or are they lost forever?
